src/vm/parser.rs

```diff
@@ -1,103 +1,165 @@
-use vm::errors::Error;
+use regex::{Regex, Captures};
+use address::c32::c32_address_decode;
+use vm::errors::{Error, InterpreterResult as Result};
 use vm::representations::SymbolicExpression;
+use vm::types::Value;
 
 #[derive(Debug)]
 pub enum LexItem {
     LeftParen,
     RightParen,
-    NameParameter(String),
-    Atom(String)
+    NamedParameter(String),
+    LiteralValue(Value),
+    Variable(String),
+    Whitespace
 }
 
-fn finish_atom(current: &mut Option<String>) -> Option<LexItem> {
-    let resp = match current {
-        &mut None => {
-            None
-        },
-        &mut Some(ref value) => {
-            if value.starts_with('#') {
-                Some(LexItem::NameParameter(value[1..].to_string()))
-            } else {
-                Some(LexItem::Atom(value.clone()))
-            }
-        },
-    };
+#[derive(Debug)]
+enum TokenType {
+    LParens, RParens, Whitespace,
+    StringLiteral, HexStringLiteral,
+    IntLiteral, QuoteLiteral,
+    Variable, NamedParameter, PrincipalLiteral
+}
 
-    *current = None;
-    resp
+struct LexMatcher {
+    matcher: Regex,
+    handler: TokenType
 }
 
-pub fn lex(input: &str) -> Result<Vec<LexItem>, Error> {
+impl LexMatcher {
+    fn new(regex_str: &str, handles: TokenType) -> LexMatcher {
+        LexMatcher {
+            matcher: Regex::new(&format!("^{}", regex_str)).unwrap(),
+            handler: handles
+        }
+    }
+}
+
+fn get_value_or_err(input: &str, captures: Captures) -> Result<String> {
+    let matched = captures.name("value").ok_or(
+        Error::ParseError("Failed to capture value from input".to_string()))?;
+    Ok(input[matched.start()..matched.end()].to_string())
+}
+
+pub fn lex(input: &str) -> Result<Vec<LexItem>> {
+    // Aaron: I'd like these to be static, but that'd require using
+    //    lazy_static (or just hand implementing that), and I'm not convinced
+    //    it's worth either (1) an extern macro, or (2) the complexity of hand implementing.
+    let lex_matchers: &[LexMatcher] = &[
+        LexMatcher::new(r##""(?P<value>((\\")|([[:ascii:]&&[^"\n\r\t]]))*)""##, TokenType::StringLiteral),
```

Comment:
What is `:ascii:` here?  Is it _printable_ ASCII characters?  Also, can we have a lexer test below for ensuring that a code snippit with non-printable characters does not parse?

Reply:
Also, the `str` type is a _unicode_ string slice.  We'll want the lexer to reject non-ASCII strings.

Reply:
Yeah -- that `:ascii:` class should actually be `:print:` for printable ascii characters.

But this lexer _will_ reject non-ASCII strings. None of these matchers will match a non-ascii character, which will cause the lexer to fail (if the lexer cannot process a given character, it exhaust all the matches and returns a ParseError)